dio/lib/src/transformers/utf8_json_transformer.dart

```diff
@@ -0,0 +1,184 @@
+    if (responseType == ResponseType.bytes) {
+      return _consolidateStream(responseBody.stream);
+    }
+
+    final isJsonContent = Transformer.isJsonMimeType(
+      responseBody.headers[Headers.contentTypeHeader]?.first,
+    );
+
+    final customResponseDecoder = options.responseDecoder;
+
+    // no custom decoder was specified for the response,
+    // and the response is json -> use the fast path decoder
+    if (isJsonContent && customResponseDecoder == null) {
+      return _fastUtf8JsonDecode(responseBody);
+    }
+    final responseBytes = await _consolidateStream(responseBody.stream);
+
+    // a custom response decoder overrides the default behavior
+    final String? decodedResponse;
+
+    if (customResponseDecoder != null) {
+      final decodeResponse = customResponseDecoder(
+        responseBytes,
+        options,
+        responseBody..stream = const Stream.empty(),
+      );
+
+      if (decodeResponse is Future) {
+        decodedResponse = await decodeResponse;
+      } else {
+        decodedResponse = decodeResponse;
+      }
+    } else {
+      decodedResponse = null;
+    }
+
+    if (isJsonContent && decodedResponse != null) {
+      // slow path decoder, since there was a custom decoder specified
+      return jsonDecode(decodedResponse);
+    } else if (decodedResponse != null) {
+      return decodedResponse;
+    } else {
+      // If the response is not JSON and no custom decoder was specified,
+      // assume it is an utf8 string
+      return utf8.decode(
+        responseBytes,
+        allowMalformed: true,
+      );
+    }
+  }
+
+  Future<Object?> _fastUtf8JsonDecode(ResponseBody responseBody) async {
+    final contentLengthHeader =
+        responseBody.headers[Headers.contentLengthHeader];
+
+    final hasContentLengthHeader =
+        contentLengthHeader != null && contentLengthHeader.isNotEmpty;
+
+    // the content length of the response, either from the content-length header
+    // of the response or the length of the eagerly decoded response bytes
+    final int contentLength;
+
+    // the eagerly decoded response bytes
+    // which is set if the content length is not specified
+    // null otherwise (we'll feed the stream directly to the decoder in that case)
+    Uint8List? responseBytes;
+
+    // if the content length is not specified, we need to consolidate the stream
+    // and count the bytes to determine if we should use an isolate
+    // otherwise we use the content length header
+    if (!hasContentLengthHeader) {
+      responseBytes = await _consolidateStream(responseBody.stream);
+      contentLength = responseBytes.length;
+    } else {
+      contentLength = int.parse(contentLengthHeader.first);
+    }
+
+    final shouldUseIsolate = !(contentLengthIsolateThreshold < 0) &&
+        contentLength >= contentLengthIsolateThreshold;
+    if (shouldUseIsolate) {
+      // we can't send the stream to the isolate, so we need to decode the response bytes first
```

Comment:
Would that lead to a memory issue compare to the previous transformer?

Reply:
no, the previous transformers always decode into a single UInt8List first.
SyncTransformer:
```dart
    final chunks = await responseBody.stream.toList();
    final responseBytes = Uint8List.fromList(chunks.expand((c) => c).toList());
```